core/src/main/scala/org/apache/spark/scheduler/TaskDescription.scala

```diff
@@ -17,27 +17,123 @@
+ *         in the TaskDescription, even though they're also in the serialized task.
+ *     (2) Because a TaskDescription is serialized and sent to an executor for each task, efficient
+ *         serialization (both in terms of serialization time and serialized buffer size) is
+ *         important. For this reason, we serialize TaskDescriptions ourselves with the
+ *         TaskDescription.encode and TaskDescription.decode methods.  This results in a smaller
+ *         serialized size because it avoids serializing unnecessary fields in the Map objects
+ *         (which can introduce significant overhead when the maps are small).
  */
 private[spark] class TaskDescription(
     val taskId: Long,
     val attemptNumber: Int,
     val executorId: String,
     val name: String,
     val index: Int,    // Index within this task's TaskSet
-    _serializedTask: ByteBuffer)
-  extends Serializable {
+    val addedFiles: Map[String, Long],
+    val addedJars: Map[String, Long],
+    val properties: Properties,
+    val serializedTask: ByteBuffer) {
 
-  // Because ByteBuffers are not serializable, wrap the task in a SerializableBuffer
-  private val buffer = new SerializableBuffer(_serializedTask)
+  override def toString: String = "TaskDescription(TID=%d, index=%d)".format(taskId, index)
+}
 
-  def serializedTask: ByteBuffer = buffer.value
+private[spark] object TaskDescription {
+  def encode(taskDescription: TaskDescription): ByteBuffer = {
+    val bytesOut = new ByteBufferOutputStream(4096)
+    val dataOut = new DataOutputStream(bytesOut)
 
-  override def toString: String = "TaskDescription(TID=%d, index=%d)".format(taskId, index)
+    dataOut.writeLong(taskDescription.taskId)
+    dataOut.writeInt(taskDescription.attemptNumber)
+    dataOut.writeUTF(taskDescription.executorId)
+    dataOut.writeUTF(taskDescription.name)
+    dataOut.writeInt(taskDescription.index)
+
+    // Write files.
+    dataOut.writeInt(taskDescription.addedFiles.size)
+    for ((name, timestamp) <- taskDescription.addedFiles) {
+      dataOut.writeUTF(name)
+      dataOut.writeLong(timestamp)
+    }
+
+    // Write jars.
+    dataOut.writeInt(taskDescription.addedJars.size)
+    for ((name, timestamp) <- taskDescription.addedJars) {
+      dataOut.writeUTF(name)
+      dataOut.writeLong(timestamp)
+    }
+
+    // Write properties.
+    dataOut.writeInt(taskDescription.properties.size())
+    taskDescription.properties.stringPropertyNames.asScala.foreach { name =>
+      dataOut.writeUTF(name)
+      dataOut.writeUTF(taskDescription.properties.getProperty(name))
+    }
+
+    // Write the task. The task is already serialized, so write it directly to the byte buffer
+    // (this requires first flushing the data output stream, so that all of the data has been
+    // written from the data output stream so the underlying ByteBufferOutputStream before
+    // we write the task).
+    dataOut.flush()
+    Utils.writeByteBuffer(taskDescription.serializedTask, bytesOut)
```

Comment:
Just opened #16462 to fix this.  I don't think this has any issues for this code path, because this is exactly the same as the old code that was previously located in Task.scala (which was moved here in this PR).